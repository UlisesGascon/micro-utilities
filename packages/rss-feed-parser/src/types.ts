export interface Item {
    title: string;
    description: string;
    pubDate: string;
    link: string;
    guid: string;
  }
  
export interface Image {
    title: string;
    url: string;
    link: string;
  }
  
export interface Channel {
    title: string;
    link: string;
    'atom:link': string;
    description: string;
    pubDate: string;
    language: string;
    docs: string;
    lastBuildDate: string;
    image: Image;
    item: Item[];
  }
  
export interface RssParseReturn {
    metadata: Omit<Channel, 'item'>;
    items: Channel;
}