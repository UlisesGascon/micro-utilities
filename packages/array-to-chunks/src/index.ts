export function chunkArray (arr:any[], chunkSize:number):any[] { 
  if(!Array.isArray(arr)) throw new Error('arr must be an array')
  if(!Number.isInteger(chunkSize) && chunkSize >= 0) throw new Error('chunkSize must be a positive integer')

  const res: any[] = []
  for (let i = 0; i < arr.length; i += chunkSize) {
      const chunk = arr.slice(i, i + chunkSize);
      res.push(chunk);
  }
  return res
}