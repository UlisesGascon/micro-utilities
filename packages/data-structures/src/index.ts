class Node {
  value: any;
  next: Node | null;

  constructor(value: any){
      this.value = value;
      this.next = null;
  }
}

export class LinkedList {
  head: Node | null;
  tail: Node | null;
  length: number;

  constructor (value:any) {
    const node = new Node(value)
    this.head = node
    this.tail = this.head
    this.length = 1
  }

  public push (value: any):LinkedList {
    const node = new Node(value)
    if(this.tail != null) {
      this.tail.next = node
    }
    this.tail = node
    this.length++
    return this
  }

  public pop () :any | undefined {
    if(this.head == null) {
      return undefined
    }

    let temp = this.head

    if(this.length === 1) {
      this.head = null
      this.tail = null
      this.length = 0
      return temp.value
    }

    while (temp.next != null) {
      temp = temp.next
    }

    this.tail = temp
    this.tail.next = null
    this.length--
    return temp.value
  }

  public shift (value:any) :any {
    if(this.head == null) {
      return undefined
    }

    const temp = this.head
    this.head = temp.next
    this.length--
    if(this.length === 0) {
      this.tail = null
    }
    return temp.value
  }

  public unshift (value:any) :LinkedList {
    const node = new Node(value)
    if(this.head == null) {
      this.head = node
      this.tail = this.head
    } else {
      node.next = this.head
      this.head = node
    }
    this.length++
    return this
  }

  public purge () :LinkedList {
    this.head = null
    this.tail = null
    this.length = 0
    return this
  }
}
