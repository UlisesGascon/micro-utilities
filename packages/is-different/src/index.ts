export function isDifferent(firstValue: any, secondValue: any): boolean {
  return JSON.stringify(firstValue) !== JSON.stringify(secondValue);
}
