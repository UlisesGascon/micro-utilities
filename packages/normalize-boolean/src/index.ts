export function normalizeBoolean (value:any):boolean {
    // Ignore non-primitive values
    if (!['boolean', 'string', 'number'].includes(typeof value)) {
      return false
    }
  
    const normalizedValue = !isNaN(parseInt(value)) ? parseInt(value) : value
  
    // Add support for string values
    if (typeof normalizedValue === 'string') {
      return ['true', 'on', 'yes', 'y'].includes(normalizedValue.toLowerCase().trim())
    }
  
    // Add support for number values
    if (typeof normalizedValue === 'number') {
      return normalizedValue === 1
    }
  
    return Boolean(value)
}