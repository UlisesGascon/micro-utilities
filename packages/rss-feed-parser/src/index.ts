import { XMLParser, XMLValidator} from 'fast-xml-parser'
import { type RssParseReturn } from './types';

const parser = new XMLParser();

export function rssParse(rssContent: string): RssParseReturn {
  const result:any = XMLValidator.validate(rssContent);
  if (result.err !== undefined) {
    throw new Error('XML is invalid');
  }
  
  const jObj = parser.parse(rssContent);

  if (jObj?.rss?.channel === undefined) {
    throw new Error('XML has not valid rss format');
  }

  const {item: items, ...metadata} = jObj.rss.channel;

  return {
    metadata,
    items: items ?? []
  }
}
