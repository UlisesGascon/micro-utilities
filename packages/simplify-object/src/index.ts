interface simplifyPayloadType {
  include?: string[]
  exclude?: string[] 
}

type OpenObjectType = Record<string, any>;

const isObject = (obj:object):boolean => {
  return typeof obj === 'object' && !Array.isArray(obj) && obj !== null
}

export function simplifyObject (baseObj:OpenObjectType, operation:simplifyPayloadType):OpenObjectType {
  if(!isObject(baseObj)) throw new Error('baseObj must be an object')
  if(!isObject(operation)) throw new Error('operation must be an object')

  if(Array.isArray(operation.include) && Array.isArray(operation.exclude)) throw new Error('operation must have either include or exclude')
  if(!Array.isArray(operation.include) && !Array.isArray(operation.exclude)) throw new Error('operation must have either include or exclude')

  if(Array.isArray(operation.include) && (operation.include.length === 0)) throw new Error('operation.include must be an array with properties')
  if(Array.isArray(operation.exclude) && (operation.exclude.length === 0)) throw new Error('operation.exclude must be an array with properties')
  
  const { include=[], exclude=[] } = operation 

  const simplifiedObject:OpenObjectType = {}

  Object.keys(baseObj).forEach(key => {
      if(((include.length > 0) && include.includes(key)) || (exclude.length > 0) && !exclude.includes(key)) {
        simplifiedObject[key] = baseObj[key]
      }
  })
  
  return simplifiedObject
}