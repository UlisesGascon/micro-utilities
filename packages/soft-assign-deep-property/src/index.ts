type OpenObjectType = Record<string, any>;

const isObject = (obj: object): boolean => {
  return typeof obj === "object" && !Array.isArray(obj) && obj !== null;
};

export function softAssign(
  obj: OpenObjectType,
  keys: string[],
  value: any
): void {
  if (!isObject(obj)) throw new Error("obj must be an object");
  if (!Array.isArray(keys) || keys.length === 0)
    throw new Error("keys must be an Array with at least one element");

  const lastKeyIndex = keys.length - 1;
  for (let i = 0; i < lastKeyIndex; ++i) {
    const key = keys[i];
    if (!(key in obj)) {
      obj[key] = {};
    }
    obj = obj[key];
  }
  obj[keys[lastKeyIndex]] =
    obj[keys[lastKeyIndex]] !== undefined ? obj[keys[lastKeyIndex]] : value;
}
