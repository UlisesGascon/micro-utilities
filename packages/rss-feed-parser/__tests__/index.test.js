import { rssParse } from '../src'
import { readFileSync } from 'fs'
import { resolve } from 'path'

const feedData = readFileSync(resolve('./__fixtures__/feed.xml'), 'utf8')

describe('rss-feed-parser', () => {
    it('Should parse a valid RSS feed', () => {
      expect(rssParse(feedData)).toMatchSnapshot()
    })

    it('Should throw an error if the feed has not the minimum required fields', () => {
      const missingChannel = '<rss></rss>';
      const badContent = '<other></other>';
      expect(() => rssParse(missingChannel)).toThrow('XML has not valid rss format');
      expect(() => rssParse(badContent)).toThrow('XML has not valid rss format');
    });

    it('Should return an empty array if the feed has no items', () => {
      const noItems = '<rss><channel><title>Test</title></channel></rss>';
      const parsed = rssParse(noItems);
      expect(parsed.items).toStrictEqual([]);
      expect(parsed.metadata).toStrictEqual({title: 'Test'})
    })

    it('Should throw an error if the feed is not valid XML', () => {
      const invalidXML = 'invalid';
      expect(() => rssParse(invalidXML)).toThrow('XML is invalid');
    })
})
