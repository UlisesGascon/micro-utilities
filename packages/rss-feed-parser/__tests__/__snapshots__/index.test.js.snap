// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rss-feed-parser Should parse a valid RSS feed 1`] = `
Object {
  "items": Array [
    Object {
      "description": "<p>Reporting on 35 Issues from 26 authors, 65 Pull Requests from 38 authors, and 13 Discussions from 12 authors.</p>",
      "guid": "https://github.com/cutenode/retro-weekly/blob/main/retros/2023-11-12.md",
      "link": "https://github.com/cutenode/retro-weekly/blob/main/retros/2023-11-12.md",
      "pubDate": "Sun, 12 Nov 2023 00:00:00 GMT",
      "title": "Retrospective for nodejs from 2023-11-12 to 2023-11-19",
    },
    Object {
      "description": "<p>Reporting on 45 Issues from 25 authors, 55 Pull Requests from 36 authors, and 9 Discussions from 9 authors.</p>",
      "guid": "https://github.com/cutenode/retro-weekly/blob/main/retros/2023-11-05.md",
      "link": "https://github.com/cutenode/retro-weekly/blob/main/retros/2023-11-05.md",
      "pubDate": "Sun, 05 Nov 2023 00:00:00 GMT",
      "title": "Retrospective for nodejs from 2023-11-05 to 2023-11-12",
    },
  ],
  "metadata": Object {
    "atom:link": "",
    "description": "Node.js Core most relevant news",
    "docs": "https://validator.w3.org/feed/docs/rss2.html",
    "image": Object {
      "link": "https://github.com/nodejs/nodejs-news-feeder",
      "title": "Node.js News",
      "url": "https://pbs.twimg.com/profile_images/1262824892535373825/BiXDFDDp_400x400.jpg",
    },
    "language": "en-us",
    "lastBuildDate": "Tue, 21 Nov 2023 07:15:00 GMT",
    "link": "https://nodejs.github.io/nodejs-news-feeder/feed.xml",
    "pubDate": "Tue, 13 Jun 2023 04:17:50 GMT",
    "title": "Node.js News",
  },
}
`;
